ctru-rs/src/services/uds.rs

```diff
@@ -0,0 +1,960 @@
+}
+
+/// Information about a network node.
+/// Ported to Rust so that Debug can be derived on it.
+#[allow(missing_docs)]
+#[doc(alias = "udsNodeInfo")]
+#[derive(Debug)]
+pub struct NodeInfo {
+    pub uds_friendcodeseed: u64,
+    pub username: String,
+    pub unk_x1c: u16,
+    pub flag: u8,
+    pub pad_x1f: u8,
+    pub node_id: u16,
+    pub pad_x22: u16,
+    pub word_x24: u32,
+}
+
+impl From<ctru_sys::udsNodeInfo> for NodeInfo {
+    fn from(value: ctru_sys::udsNodeInfo) -> Self {
+        unsafe {
+            Self {
+                uds_friendcodeseed: value.uds_friendcodeseed,
+                username: String::from_utf16_lossy(
+                    &value.__bindgen_anon_1.__bindgen_anon_1.username,
+                ),
+                unk_x1c: value.__bindgen_anon_1.__bindgen_anon_1.unk_x1c,
+                flag: value.__bindgen_anon_1.__bindgen_anon_1.flag,
+                pad_x1f: value.__bindgen_anon_1.__bindgen_anon_1.pad_x1f,
+                node_id: value.NetworkNodeID,
+                pad_x22: value.pad_x22,
+                word_x24: value.word_x24,
+            }
+        }
+    }
+}
+
+/// Information returned from scanning for networks.
+/// Ported to Rust so that Debug can be derived on it.
+#[allow(missing_docs)]
+#[doc(alias = "udsNetworkScanInfo")]
+#[derive(Debug)]
+pub struct NetworkScanInfo {
+    pub datareply_entry: ctru_sys::nwmBeaconDataReplyEntry,
+    pub network: ctru_sys::udsNetworkStruct,
+    pub nodes: [Option<NodeInfo>; 16],
+}
+
+impl From<ctru_sys::udsNetworkScanInfo> for NetworkScanInfo {
+    fn from(value: ctru_sys::udsNetworkScanInfo) -> Self {
+        Self {
+            datareply_entry: value.datareply_entry,
+            network: value.network,
+            nodes: value.nodes.map(|n| {
+                if n.uds_friendcodeseed != 0 {
+                    Some(n.into())
+                } else {
+                    None
+                }
+            }),
+        }
+    }
+}
+
+/// Status of the service handle.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum ServiceStatus {
+    /// Not connected to or hosting a network.
+    Disconnected,
+    /// Connected to a network.
+    Client,
+    /// Hosting a network.
+    Server,
+}
+
+/// Handle to the UDS service.
+pub struct Uds {
+    _service_handler: ServiceReference,
+    context: Option<ctru_sys::udsBindContext>,
+    network: Option<ctru_sys::udsNetworkStruct>,
+}
+
+static UDS_ACTIVE: Mutex<()> = Mutex::new(());
+
+impl Uds {
+    /// Size of one frame.
+    const RECV_FRAME_SIZE: usize = ctru_sys::UDS_DATAFRAME_MAXSIZE as usize;
+
+    /// Size of receive buffer; max frame size * 8.
+    const RECV_BUF_SIZE: u32 = ctru_sys::UDS_DEFAULT_RECVBUFSIZE;
+
+    /// Shared memory size; must be slightly larger
+    /// than `RECV_BUF_SIZE`.
+    const SHAREDMEM_SIZE: usize = 0x3000;
+
+    /// Buffer used while scanning for networks.
+    /// This value is taken from the devkitPRO example.
+    const SCAN_BUF_SIZE: usize = 0x4000;
```

Comment:
This buffer has to hold all of the info for beacons found during the scanning process, which can essentially be arbitrarily large depending on how many beacons there are; the devkitPRO example uses 16KB.
However, looking at the example code, I realise now I was a bit silly, because the buffer can absolutely just be allocated on the heap once when the service is initialised and reused for each scan. Is this acceptable?

Reply:
I guess so, it would surely help mitigate memory fragmentation. Now I imagine a room filled with 200 Nintendo 3DS trying to download play and the chaos that would ensue 🥸